src/NUnitFramework/framework/Internal/Execution/WorkItem.cs

```diff
@@ -211,42 +248,27 @@ public virtual void Execute()
             // execution. Currently, test cases are always run sequentially,
             // so this continues to apply fairly generally.
 
-            var ownThreadReason = OwnThreadReason.NotNeeded;
-
-#if !PORTABLE
-            if (Test.RequiresThread)
-                ownThreadReason |= OwnThreadReason.RequiresThread;
-            if (timeout > 0 && Test is TestMethod)
-                ownThreadReason |= OwnThreadReason.Timeout;
-#if !NETSTANDARD1_6
-            CurrentApartment = Thread.CurrentThread.GetApartmentState();
-            if (CurrentApartment != TargetApartment && TargetApartment != ApartmentState.Unknown)
-                ownThreadReason |= OwnThreadReason.DifferentApartment;
-#endif
-#endif
-
-            if (ownThreadReason == OwnThreadReason.NotNeeded)
-                RunTest();
-            else if (Context.IsSingleThreaded)
+            if (Test.RequiresThread || timeout > 0 || targetApartment != CurrentApartment)
             {
-                var msg = "Test is not runnable in single-threaded context. " + ownThreadReason;
-                log.Error(msg);
-                Result.SetResult(ResultState.NotRunnable, msg);
-                WorkItemComplete();
+                if (!Context.IsSingleThreaded)
+                    RunTestOnOwnThread(timeout, targetApartment);
+                else
+                {
+                    var msg = timeout > 0
+                        ? "TimeoutAttribute may not be specified on a test within a single-threaded fixture."
+                        : Test.RequiresThread
+                            ? "RequiresThreadAttribute may not be specified on a test withihn a single-SingleThreadedAttribute fixture."
+                            : "Tests in a single-threaded fixture may not specify a different apartment";
```

Comment:
I go back and forth in situations like this. Maybe it would look better with the ternary operator using a different layout, like...

```C#
var msg = condition1 ? message1 :
          condition2 ? message2 :
          condition3 ? message3 : null;
```